sumpy/qbx.py

```diff
@@ -62,45 +62,67 @@ def stringify_expn_index(i):
             return str(i)
 
 
-def expand(expansion_nr, sac, expansion, avec, bvec):
-    rscale = sym.Symbol("rscale")
-
-    coefficients = expansion.coefficients_from_source(expansion.kernel,
-                                                      avec, bvec, rscale)
-
-    assigned_coeffs = [
-            sym.Symbol(
-                    sac.assign_unique("expn%dcoeff%s" % (
-                        expansion_nr, stringify_expn_index(i)),
-                        coefficients[expansion.get_storage_index(i)]))
-            for i in expansion.get_coefficient_identifiers()]
-
-    return sac.assign_unique("expn%d_result" % expansion_nr,
-            expansion.evaluate(assigned_coeffs, bvec, rscale))
-
-
 # {{{ layer potential computation
 
 # {{{ base class
 
 class LayerPotentialBase(KernelComputation, KernelCacheWrapper):
-    def __init__(self, ctx, expansions, strength_usage=None,
-            value_dtypes=None,
-            name=None, device=None):
-        KernelComputation.__init__(self, ctx, expansions, strength_usage,
-                value_dtypes,
-                name, device)
+    def __init__(self, ctx, expansion, strength_usage=None,
+            value_dtypes=None, name=None, device=None,
+            source_kernels=None, target_kernels=None):
 
+        from sumpy.kernel import SourceDerivativeRemover, TargetDerivativeRemover
         from pytools import single_valued
-        self.dim = single_valued(knl.dim for knl in self.expansions)
+        sdr = SourceDerivativeRemover()
+        tdr = TargetDerivativeRemover()
+        if source_kernels is None and target_kernels is None:
+            from warnings import warn
+            warn("A list for the second argument of the constructor is deprecated."
+                "Use a single expansion and pass source_kernels and target_kernels",
+                DeprecationWarning, stacklevel=2)
+            source_kernel = single_valued(tdr(exp.kernel) for exp in expansion)
+            base_kernel = sdr(source_kernel)
+            target_kernels = tuple(sdr(exp.kernel) for exp in expansion)
+            source_kernels = (source_kernel,)
+            expansion = expansion[0].copy(kernel=base_kernel)
+
+        KernelComputation.__init__(self, ctx=ctx, target_kernels=target_kernels,
+                strength_usage=strength_usage, source_kernels=source_kernels,
+                value_dtypes=value_dtypes, name=name, device=device)
+
+        self.dim = single_valued(knl.dim for knl in self.target_kernels)
+        self.expansion = expansion
 
     def get_cache_key(self):
-        return (type(self).__name__, tuple(self.kernels),
-                tuple(self.strength_usage), tuple(self.value_dtypes))
+        return (type(self).__name__, self.expansion, tuple(self.target_kernels),
+                tuple(self.source_kernels), tuple(self.strength_usage),
+                tuple(self.value_dtypes))
```

Comment:
Add device?

Reply:
Fails with ` unsupported type for persistent hash keying: <class 'pyopencl._cl.Device'>`